File: PRACTICA1/RUBY/Deepspace/SpaceCity.rb
#encoding:utf-8
require './SpaceCityToUI'
module Deepspace
# @author Lucía Salamanca López, Celia Arias Martínez
class SpaceCity < SpaceStation
    def initialize(base, rest)   #base es SpaceStation original y rest es Array de SpaceStation (las demás estaciones)
        copy(base)  
        @base = base
        @collaborators = rest #array de SpaceStation, queremos el puntero
    end 

    def collaborators 
        @collaborators
    end

    
    def fire 
        suma = super  #creo que es lo mismo que super.fire
        @collaborators.each{ |f| 
            suma+=f.fire
        }
        return suma
    end 

    
    def protection
        suma = super  #creo que es lo mismo que super.protection
        @collaborators.each{ |pr|
            suma+=pr.protection
        }
        return suma
    end 

    def setLoot(loot)  
        super
        return Transformation::NOTRANSFORM
    end

    def getUIversion
        SpaceCityToUI.new(self)
    end

    def to_s  
        "La estación espacial original es \n" + super + "\n y las estaciones espaciales colaboradoras son\n #{@collaborators.join("\n ")}."
    end
end 

end

File: PRACTICA1/JAVA/deepspace/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author Celia Arias Martínez, Lucía Salamanca López
 */
public class SpaceCity extends SpaceStation{
    
    private ArrayList<SpaceStation> collaborators=new ArrayList<>();
    private SpaceStation base;
    
    
    public SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        for (SpaceStation s : rest) {
         collaborators.add(new SpaceStation(s));
        }
        base = base;
    }
    
    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    @Override
    public float fire(){
        float factor = super.fire();
        for (int i=0; i<collaborators.size();i++)
            factor+=collaborators.get(i).fire();
        
        return factor;
    }
    
    @Override
    public float protection(){
        float factor = super.protection();
        for (int i=0; i<collaborators.size();i++)
            factor+=collaborators.get(i).protection();
        
        return factor;
        
        
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        Transformation trans = Transformation.NOTRANSFORM;
        return trans;
    }
    
    @Override
    public SpaceCityToUI getUIversion(){
         return new SpaceCityToUI(this);
    }
    
}
